File: v1/haskell/say/Say.hs
module Say (inEnglish) where

import Data.Maybe (catMaybes)
import qualified Data.Vector as V

inEnglish :: Integral a => a -> Maybe String
inEnglish n 
    | n < 0 || n >= 1000000000000 = Nothing
    | n == 0 = Just "zero"
    | otherwise = Just $ unwords $ catMaybes [ billions billionsCount
                                             , millions millionsCount
                                             , thousands thousandsCount
                                             , hundreds remainder]
     where
      (billionsCount, millionsCount, thousandsCount, remainder) = parts n

parts :: Integral a => a -> (Int, Int, Int, Int)
parts n = ( fromIntegral billionsCount
          , fromIntegral millionsCount
          , fromIntegral thousandsCount
          , fromIntegral remainder) where
  (billionsCount, billionsRemainder) = n `quotRem` 1000000000
  (millionsCount, millionsRemainder) = billionsRemainder `quotRem` 1000000
  (thousandsCount, remainder) = millionsRemainder `quotRem` 1000

base :: Int -> Maybe String
base n = values V.!? (n - 1) where
    values = V.fromList [ "one"
                        , "two"
                        , "three"
                        , "four"
                        , "five"
                        , "six"
                        , "seven"
                        , "eight"
                        , "nine"
                        , "ten"
                        , "eleven"
                        , "twelve"
                        , "thirteen"
                        , "fourteen"
                        , "fifteen"
                        , "sixteen"
                        , "seventeen"
                        , "eighteen"
                        , "nineteen"]

tens :: Int -> Maybe String
tens n = 
    if n < 20 then base n
    else Just $ countStr ++ maybe "" ("-" ++) (base remainder) where
        (count, remainder) = n `quotRem` 10
        countStr = values V.! (count - 2) 
        values = V.fromList [ "twenty"
                            , "thirty"
                            , "forty"
                            , "fifty"
                            , "sixty"
                            , "seventy"
                            , "eighty"
                            , "ninety"]

hundreds :: Int -> Maybe String
hundreds n = 
    if n < 100 then tens n
    else fmap (++ " hundred" ++ tensStr) (base count) where
      tensStr = maybe "" (" " ++) (tens remainder)
      (count, remainder) = n `quotRem` 100

thousands :: Int -> Maybe String
thousands = chunk "thousand"

millions :: Int -> Maybe String
millions = chunk "million"

billions :: Int -> Maybe String
billions = chunk "billion"

chunk :: String -> Int -> Maybe String
chunk str = fmap (++ " " ++ str) . hundreds

File: v1/ecmascript/say/say.js
export default class Say {
    inEnglish(number) {
        if (number < 0 || number > 999999999999) 
            throw new Error('Number must be between 0 and 999,999,999,999.');

        if (number === 0)
            return 'zero';

        return this.parts(number).join(' ');
    }

    parts(number) {
        const [billionsCount, millionsCount, thousandsCount, remainder] = this.counts(number);
        
        return [this.billions(billionsCount), 
                this.millions(millionsCount),
                this.thousands(thousandsCount),
                this.hundreds(remainder)]
                .filter(x => x !== null);
    }

    bases(number) {
        const values = [
            "one",
            "two",
            "three",
            "four",
            "five",
            "six",
            "seven",
            "eight",
            "nine",
            "ten",
            "eleven",
            "twelve",
            "thirteen",
            "fourteen",
            "fifteen",
            "sixteen",
            "seventeen",
            "eighteen",
            "nineteen"
        ];

        return number > 0 && number <= values.length ? values[number - 1] : null;
    }

    tens(number) {
        if (number < 20)
            return this.bases(number);

        const values = [
            "twenty",
            "thirty",
            "forty",
            "fifty",
            "sixty",
            "seventy",
            "eighty",
            "ninety"
        ];

        const count = Math.floor(number / 10);
        const remainder = number % 10;
        const bases = this.bases(remainder);

        const countStr = values[count - 2];
        const basesStr = bases == null ? "" : `-${bases}`;

        return `${countStr}${basesStr}`;
    }

    hundreds(number){
        if (number < 100)
            return this.tens(number);
        
        const count = Math.floor(number / 100);
        const remainder = number % 100;
        const bases = this.bases(count);
        const tens = this.tens(remainder);
        const tensStr = tens == null ? "" : ` ${tens}`;
        
        return `${bases} hundred${tensStr}`;
    }

    chunk(str, number) {
        const hundreds = this.hundreds(number);
        return hundreds == null ? null : `${hundreds} ${str}`;
    }

    thousands(number) {
        return this.chunk("thousand", number);
    }

    millions(number) {
        return this.chunk("million", number);
    }

    billions(number) {
        return this.chunk("billion", number);
    }

    counts(number) {
        const billionsCount = Math.floor(number / 1000000000);
        const billionsRemainder = number % 1000000000;

        const millionsCount = Math.floor(billionsRemainder / 1000000);
        const millionsRemainder = billionsRemainder % 1000000;

        const thousandsCount = Math.floor(millionsRemainder / 1000);
        const thousandsRemainder = millionsRemainder % 1000;

        return [billionsCount, millionsCount, thousandsCount, thousandsRemainder];
    }
}